nntrainer/graph/network_graph.cpp

```diff
@@ -0,0 +1,424 @@
+// SPDX-License-Identifier: Apache-2.0
+/**
+ * Copyright (C) 2020 Jijoong Moon <jijoong.moon@samsung.com>
+ *
+ * @file    network_graph.h
+ * @date    19 Oct 2020
+ * @see     https://github.com/nnstreamer/nntrainer
+ * @author  Jijoong Moon <jijoong.moon@samsung.com>
+ * @bug     No known bugs except for NYI items
+ * @brief   This is Network Graph Class for Neural Network
+ *
+ */
+
+#include <layer_factory.h>
+#include <network_graph.h>
+
+#include <nntrainer_error.h>
+#include <nntrainer_log.h>
+#include <parse_util.h>
+
+namespace nntrainer {
+
+void NetworkGraph::addEdge(unsigned int ith, LayerNode node) {
+  if (ith > num_node - 1)
+    throw std::invalid_argument("Exceed total number of layer");
+
+  adj[ith].push_back(node);
+}
+
+void NetworkGraph::addLayerNode(std::shared_ptr<Layer> layer) {
+  std::list<LayerNode> l;
+  LayerNode *node = new LayerNode();
+
+  node->layer = layer;
+  node->index = num_node;
+
+  l.assign(1, *node);
+  adj.push_back(l);
+  num_node++;
+}
+
+void NetworkGraph::topologicalSortUtil(unsigned int ith, bool visited[],
+                                       std::stack<LayerNode> &Stack) {
+  visited[ith] = true;
+
+  std::list<LayerNode>::iterator i;
+  for (i = adj[ith].begin(); i != adj[ith].end(); ++i) {
+    if (!visited[(*i).index])
+      topologicalSortUtil((*i).index, visited, Stack);
+  }
+
+  Stack.push(getLayerNode(ith));
+}
+
+LayerNode &NetworkGraph::getLayerNode(unsigned int ith) {
+
+  std::list<LayerNode>::iterator iter;
+  for (unsigned int i = 0; i < num_node; ++i) {
+    iter = adj[i].begin();
+    if ((*iter).index == ith) {
+      return *iter;
+    }
+  }
+
+  throw std::invalid_argument("Cannot find Layer");
+}
+
+void NetworkGraph::topologicalSort() {
+  std::stack<LayerNode> Stack;
+
+  bool *visited = new bool[num_node];
```

Comment:
lets use `vector<bool>` or add it as a TODO here.

Reply:
done.